test/object-store/sync/flx_role_change.cpp

```diff
@@ -857,4 +857,299 @@ TEST_CASE("flx: role changes during bootstrap complete successfully", "[sync][fl
+                // Only expecting a client reset error to be reported
+                if (error.status != ErrorCodes::SyncClientResetRequired)
+                    FAIL(util::format("Unexpected error received by error handler: %1", error.status));
+            };
+        };
+
+        // Make sure the rules are reset back to the original value (only manager & director records)
+        update_role(default_rule, {{"role", {{"$in", {"manager", "director"}}}}});
+        logger->debug("ROLE CHANGE: Initial rule definitions: %1", default_rule);
+        REQUIRE(app_session.admin_api.update_default_rule(app_session.server_app_id, default_rule));
+
+        auto config_1 = harness->make_test_file();
+        auto&& [reset_future, reset_handler] = reset_utils::make_client_reset_handler();
+        config_1.sync_config->notify_after_client_reset = std::move(reset_handler);
+        config_1.sync_config->client_resync_mode = ClientResyncMode::Recover;
+        setup_config_callbacks(config_1);
+
+        // Wait for download, upload and advance - uses a bool flag instead of REQUIRE()
+        // to address TSAN failures around the catch2 library.
+        auto wait_for_update = [](SharedRealm realm) {
+            bool success = !wait_for_download(*realm);
+            success = success && !wait_for_upload(*realm);
+            if (!success)
+                FAIL("Failed to update realm");
+            wait_for_advance(*realm);
+        };
+
+        auto realm_1 = Realm::get_shared_realm(config_1);
+        {
+            // Set up a new bootstrap while offline
+            auto table = realm_1->read_group().get_table("class_Person");
+            auto new_subs = realm_1->get_latest_subscription_set().make_mutable_copy();
+            new_subs.clear();
+            new_subs.insert_or_assign(Query(table));
+            auto subs = new_subs.commit();
+            subs.get_state_change_notification(sync::SubscriptionSet::State::Complete).get();
+            wait_for_update(realm_1);
+            verify_records(realm_1, 0, params.num_mgrs, params.num_dirs);
+        }
+        // The test will update the rule to change access from only manager and director records
+        // to only the employee records while a client reset is in progress.
+        update_role(default_rule, {{"role", "employee"}});
+        // Force a client reset to occur the next time the session connects
+        reset_utils::trigger_client_reset(app_session, realm_1);
+        bool skip_role_change_check = false;
+
+        // Set the state where the role change will be triggered
+        auto set_expected_role_state = [&](ClientResetTestState state, bool skip_role_check = false) {
+            client_reset_state.transition_with([&](ClientResetTestState) {
+                update_role_state = state;
+                // If the role change check is skipped, the test will not look for the role change error
```

Comment:
Do we want to track the role change bootstrap (where it makes sense) instead? we only seem to track the bootstrap of the client reset session.

Reply:
Based on the differing order of hook events depending on where the role change occurs, it is difficult to detect which bootstrap is the role change bootstrap (and sometimes it happens on the fresh realm session, while other times it happens on the primary session). Instead, the test is looking for the 200 error, which should be triggered by the role change, but it is sometimes hard to catch this using the hook event callback, especially when the session is being torn down.

Reply:
The tracking of the client reset fresh realm bootstrap is so we can trigger the role change at different points along the client reset chain of events.